File: node_engine/src/checks/EmptyFormLabelCheck.js
const { getAttributeValue } = require("./AttributeParser");
const { getLineNumber } = require("./TextUtilities");
const {
  collectElementIds,
  hasAriaLabel,
  hasValidAriaLabelledBy,
  hasTitle,
  hasTextContent
} = require("./AccessibleNameUtilities");

const labelRegex = /<label(?<attrs>[^>]*)>(?<content>[\s\S]*?)<\/label>/gi;

const EmptyFormLabelCheck = {
  id: "empty-form-label",
  applicableKinds: ["html", "htm", "cshtml", "razor"],
  run(context, rule) {
    const issues = [];
    const elementIds = collectElementIds(context.content);

    for (const match of context.content.matchAll(labelRegex)) {
      const attrs = match.groups?.attrs ?? "";
      const content = match.groups?.content ?? "";

      if (hasTextContent(content)) {
        continue;
      }

      const ariaLabel = getAttributeValue(attrs, "aria-label");
      if (hasAriaLabel(ariaLabel)) {
        continue;
      }

      const labelledBy = getAttributeValue(attrs, "aria-labelledby");
      if (hasValidAriaLabelledBy(labelledBy, elementIds)) {
        continue;
      }

      const title = getAttributeValue(attrs, "title");
      if (hasTitle(title)) {
        continue;
      }

      const line = getLineNumber(context.content, match.index);
      issues.push({
        ruleId: rule.id,
        checkId: EmptyFormLabelCheck.id,
        filePath: context.filePath,
        line,
        message: "Form label has no readable text.",
        evidence: match[0]
      });
    }

    return issues;
  }
};

module.exports = { EmptyFormLabelCheck };

File: src/Scanner.Core/Checks/EmptyFormLabelCheck.cs
using System.Text.RegularExpressions;
using Scanner.Core.Rules;

namespace Scanner.Core.Checks;

/// <summary>
/// Checks for form labels that lack readable text content.
/// </summary>
public sealed class EmptyFormLabelCheck : ICheck
{
    private static readonly Regex LabelRegex = new("<label(?<attrs>[^>]*)>(?<content>.*?)</label>", RegexOptions.IgnoreCase | RegexOptions.Singleline | RegexOptions.Compiled);

    /// <inheritdoc />
    public string Id => "empty-form-label";

    /// <inheritdoc />
    public IReadOnlyCollection<string> ApplicableKinds { get; } = new[] { "html", "htm", "cshtml", "razor" };

    /// <inheritdoc />
    public IEnumerable<Issue> Run(CheckContext context, RuleDefinition rule)
    {
        var issues = new List<Issue>();
        var elementIds = AccessibleNameUtilities.CollectElementIds(context.Content);

        foreach (Match match in LabelRegex.Matches(context.Content))
        {
            var attrs = match.Groups["attrs"].Value;
            var content = match.Groups["content"].Value;

            if (AccessibleNameUtilities.HasTextContent(content))
            {
                continue;
            }

            var ariaLabel = AttributeParser.GetAttributeValue(attrs, "aria-label");
            if (AccessibleNameUtilities.HasAriaLabel(ariaLabel))
            {
                continue;
            }

            var ariaLabelledBy = AttributeParser.GetAttributeValue(attrs, "aria-labelledby");
            if (AccessibleNameUtilities.HasValidAriaLabelledBy(ariaLabelledBy, elementIds))
            {
                continue;
            }

            var title = AttributeParser.GetAttributeValue(attrs, "title");
            if (AccessibleNameUtilities.HasTitle(title))
            {
                continue;
            }

            var line = TextUtilities.GetLineNumber(context.Content, match.Index);
            issues.Add(new Issue(rule.Id, Id, context.FilePath, line, "Form label has no readable text.", match.Value));
        }

        return issues;
    }
}
